coderd/batchstats/batcher.go

```diff
@@ -0,0 +1,255 @@
+	}
+
+	if b.batchSize == 0 {
+		b.batchSize = DefaultBatchSize
+	}
+
+	b.connectionsByProto = make([]map[string]int64, 0, b.batchSize)
+	b.buf = database.InsertWorkspaceAgentStatsParams{
+		ID:                          make([]uuid.UUID, 0, b.batchSize),
+		CreatedAt:                   make([]time.Time, 0, b.batchSize),
+		UserID:                      make([]uuid.UUID, 0, b.batchSize),
+		WorkspaceID:                 make([]uuid.UUID, 0, b.batchSize),
+		TemplateID:                  make([]uuid.UUID, 0, b.batchSize),
+		AgentID:                     make([]uuid.UUID, 0, b.batchSize),
+		ConnectionsByProto:          json.RawMessage("[]"),
+		ConnectionCount:             make([]int64, 0, b.batchSize),
+		RxPackets:                   make([]int64, 0, b.batchSize),
+		RxBytes:                     make([]int64, 0, b.batchSize),
+		TxPackets:                   make([]int64, 0, b.batchSize),
+		TxBytes:                     make([]int64, 0, b.batchSize),
+		SessionCountVSCode:          make([]int64, 0, b.batchSize),
+		SessionCountJetBrains:       make([]int64, 0, b.batchSize),
+		SessionCountReconnectingPTY: make([]int64, 0, b.batchSize),
+		SessionCountSSH:             make([]int64, 0, b.batchSize),
+		ConnectionMedianLatencyMS:   make([]float64, 0, b.batchSize),
+	}
+
+	return b, nil
+}
+
+// Add adds a stat to the batcher for the given workspace and agent.
+func (b *Batcher) Add(
+	agentID uuid.UUID,
+	templateID uuid.UUID,
+	userID uuid.UUID,
+	workspaceID uuid.UUID,
+	st agentsdk.Stats,
+) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	now := database.Now()
+
+	b.buf.ID = append(b.buf.ID, uuid.New())
+	b.buf.CreatedAt = append(b.buf.CreatedAt, now)
+	b.buf.AgentID = append(b.buf.AgentID, agentID)
+	b.buf.UserID = append(b.buf.UserID, userID)
+	b.buf.TemplateID = append(b.buf.TemplateID, templateID)
+	b.buf.WorkspaceID = append(b.buf.WorkspaceID, workspaceID)
+
+	// Store the connections by proto separately as it's a jsonb field. We marshal on flush.
+	// b.buf.ConnectionsByProto = append(b.buf.ConnectionsByProto, st.ConnectionsByProto)
+	b.connectionsByProto = append(b.connectionsByProto, st.ConnectionsByProto)
+
+	b.buf.ConnectionCount = append(b.buf.ConnectionCount, st.ConnectionCount)
+	b.buf.RxPackets = append(b.buf.RxPackets, st.RxPackets)
+	b.buf.RxBytes = append(b.buf.RxBytes, st.RxBytes)
+	b.buf.TxPackets = append(b.buf.TxPackets, st.TxPackets)
+	b.buf.TxBytes = append(b.buf.TxBytes, st.TxBytes)
+	b.buf.SessionCountVSCode = append(b.buf.SessionCountVSCode, st.SessionCountVSCode)
+	b.buf.SessionCountJetBrains = append(b.buf.SessionCountJetBrains, st.SessionCountJetBrains)
+	b.buf.SessionCountReconnectingPTY = append(b.buf.SessionCountReconnectingPTY, st.SessionCountReconnectingPTY)
+	b.buf.SessionCountSSH = append(b.buf.SessionCountSSH, st.SessionCountSSH)
+	b.buf.ConnectionMedianLatencyMS = append(b.buf.ConnectionMedianLatencyMS, st.ConnectionMedianLatencyMS)
+
+	// If the buffer is full, signal the flusher to flush immediately.
```

Comment:
nit: do you think that we should flush before the buffer is full? Let's say 80% of occupancy?

Reply:
What would the advantage of this be? If you're trying to avoid blocking during flush, blocking will happen regardless of when we flush. We could make a temporary copy of the buffer while we flush to avoid that, but batch-inserting 1000 entries happens fairly quickly from what I can tell.

Reply:
We had a conversation about this.

We came up with passing the existing buffer to the lever and resetting the buffer. So that way the buffer is new for the next `Add` and the previous data is pending the db query.

Psuedo code:
```golang
	if len(b.buf.ID) == cap(b.buf.ID) {
		select{
		case b.flushLever <- struct{}{}:
		default:
			b.log.Error(context.Background(), "this should never happen, dropping agent stats :(")
		}
		b.flushLever <- struct{}{}
		b.flushLever <- &b.buf
		b.buf = make(database.InsertWorkspaceAgentStatsParams, 0, b.batchSize)
	}
```

Some additional nits:
- We should probably use the constant instead of `cap` for comparison since `cap` can grow. `len(b.buf.ID) >= b.batchSize`
- We should use a select and throw an error log if we are backing up. We should drop the stats, as if we are backing up, there is nothing we can do except log it anyway.

Reply:
It ended up being simpler to do an early flush. However, I thought about this some more and I think using a buffered channel would also resolve this issue. This is a possible later enhancement.